kardia/params.go

```diff
@@ -47,3 +51,152 @@ func (ec *Client) GetMaxProposers(ctx context.Context) (int64, error) {
+	var result struct {
+		VoteYes     *big.Int
+		VoteNo      *big.Int
+		VoteAbstain *big.Int
+		ParamKeys   []uint8
+		ParamValues []*big.Int
+	}
+	// unpack result
+	err = ec.paramsUtil.Abi.UnpackIntoInterface(&result, "getProposalDetails", res)
+	if err != nil {
+		ec.lgr.Error("Error unpacking proposal", zap.String("ID", proposalId.String()), zap.Error(err))
+		return nil, err
+	}
+	metadata, err := ec.getProposalMetadata(ctx, proposalId)
+	if err != nil {
+		return nil, err
+	}
+	detail := &types.ProposalDetail{
+		ProposalMetadata: *metadata,
+		VoteYes:          result.VoteYes.Uint64(),
+		VoteNo:           result.VoteNo.Uint64(),
+		VoteAbstain:      result.VoteAbstain.Uint64(),
+		Params:           make(map[string]string),
+	}
+	for i, key := range result.ParamKeys {
+		detail.Params[cfg.ParamKeys[key]] = result.ParamValues[i].String()
+	}
+	return detail, nil
+}
+
+// GetProposalMetadata returns metadata of a proposal by ID
+func (ec *Client) getProposalMetadata(ctx context.Context, proposalId *big.Int) (*types.ProposalMetadata, error) {
+	payload, err := ec.paramsUtil.Abi.Pack("proposals", proposalId)
+	if err != nil {
+		return nil, err
+	}
+	res, err := ec.KardiaCall(ctx, contructCallArgs(ec.paramsUtil.ContractAddress.Hex(), payload))
+	if err != nil {
+		return nil, err
+	}
+
+	var result struct {
+		Proposer  common.Address
+		StartTime *big.Int
+		EndTime   *big.Int
+		Deposit   *big.Int
+		Status    uint8
+	}
+	// unpack result
+	err = ec.paramsUtil.Abi.UnpackIntoInterface(&result, "proposals", res)
+	if err != nil {
+		ec.lgr.Error("Error unpacking proposal metadata", zap.String("ID", proposalId.String()), zap.Error(err))
+		return nil, err
+	}
+	return &types.ProposalMetadata{
+		ID:        proposalId.Uint64(),
+		Proposer:  result.Proposer.String(),
+		StartTime: result.StartTime.Uint64(),
+		EndTime:   result.EndTime.Uint64(),
+		Deposit:   result.Deposit.String(),
+		Status:    result.Status,
+	}, nil
+}
+
+// GetTotalProposals returns total number of proposals
+func (ec *Client) GetTotalProposals(ctx context.Context) (*big.Int, error) {
+	payload, err := ec.paramsUtil.Abi.Pack("allProposal")
+	if err != nil {
+		return nil, err
+	}
+	res, err := ec.KardiaCall(ctx, contructCallArgs(ec.paramsUtil.ContractAddress.Hex(), payload))
+	if err != nil {
+		return nil, err
+	}
+
+	var result struct {
+		Total *big.Int
+	}
+	// unpack result
+	err = ec.paramsUtil.Abi.UnpackIntoInterface(&result, "allProposal", res)
+	if err != nil {
+		ec.lgr.Error("Error unpacking total proposals", zap.Error(err))
+		return nil, err
+	}
+	return result.Total, nil
+}
+
+// GetProposals returns list of proposals
+func (ec *Client) GetProposals(ctx context.Context, pagination *types.Pagination) ([]*types.ProposalMetadata, uint64, error) {
+	one := big.NewInt(1)
+	start := new(big.Int).SetInt64(int64(pagination.Skip))
+	end := new(big.Int).SetInt64(int64(pagination.Limit))
+	total, err := ec.GetTotalProposals(ctx)
+	if err != nil {
+		return nil, 0 , err
```

Comment:
🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>File is not `gofmt`-ed with `-s` (gofmt)